src/gui/dialogs/editor/edit_pbl.cpp

```diff
@@ -394,13 +395,22 @@ void editor_edit_pbl::select_icon_file()
 	if(dlg.show()) {
 		std::string path = dlg.path();
 		if(path.find(filesystem::get_core_images_dir()+"/icons/") == 0) {
-			std::string icon = path.substr(filesystem::get_core_images_dir().length());
+			std::string icon = path.substr(filesystem::get_core_images_dir().length()+1);
 			// setting this programmatically doesn't seem to trigger connect_signal_notify_modified()
 			find_widget<text_box>(get_window(), "icon", false).set_value(icon);
 			find_widget<drawing>(get_window(), "preview", false).set_label(icon);
 		} else {
-			// TODO: convert image file to data uri
-			// filesystem::scoped_istream nonmainline_icon = filesystem::istream_file(path);
+			// TODO: needing two vectors here is dumb
+			std::vector<char> t = filesystem::read_file_binary(path);
+			std::vector<unsigned char> t1;
+			for(char c : t) {
+				t1.emplace_back(c);
+			}
+			utils::byte_string_view v = {t1.data(), t1.size()};
+			std::string name = filesystem::base_name(path);
+			std::string img = "data:image/"+name.substr(name.find(".")+1)+";base64,"+base64::encode(v);
```

Comment:
Just checking, is there not a function already that does this?

Reply:
Not that I've found, at least. Even just grepping the src folder for the `data:image` part of the string doesn't find anything. If there was something that did it already, I would guess that the DataURI wiki page wouldn't say to find an external site to use to convert images.

Reply:
And just checking, but `base64::encode` is using the regular alphabet, not the so-called "web-safe" alphabet, right? That is, it contains + and / characters in the output rather than - and _

Reply:
Which is actually a problem since the pluses get removed.

Reply:
Huh? Isn't it wrapped in quotes so that doesn't happen?

Reply:
For whatever reason it doesn't end up getting wrapped in double quotes. The `<<` and `>>` does fix the issue though.

Reply:
An example of the output:
[_server.txt](https://github.com/wesnoth/wesnoth/files/12034983/_server.txt)

Reply:
Well, it's definitely wrong. The WML should not be serialized like that. If there are literal `<<` and `>>` in the config, it should be serialized so that there are literal `<<` and `>>` when read back in. If there are literal `+` in the config, it should be serialized so that there are literal `+` when read back in.

So, I think you've uncovered a rather nasty bug in the WML serializer. Perhaps it only quotes fields if they have whitespace in them? That would be my first guess, at least.

Reply:
Nope - adding a space to an attribute still results in no double quotes being present. ie:
`title = eee g`

Reply:
There is something very strange about this, because when I use `wml.tostring` everything is surrounded in quotes. How is what you're doing so different?

Reply:
Okay, yes, don't write `debug()` to file. That's not expected to work. And if a streaming operator uses `debug()` then I think I'd call that a bug. The streaming operator should either use `config_writer` or (more likely) simply not exist (forcing you to decide whether to explicitly call `debug()` or use a `config_writer`).